src/main/java/frc/robot/subsystems/swerve/PhoenixOdometryThread.java

```diff
@@ -136,7 +137,12 @@ public void run() {
           journal.add(
               new Samples(timestampFor(signals), Maps.asMap(signals, s -> s.getValueAsDouble())));
         } else {
-          // System.out.println("Odo thread error: " + status.toString());
+          journal.add(
+              new Samples(
+                  timestampFor(signals),
+                  signals.stream()
+                      .filter(s -> s.getStatus().equals(StatusCode.OK))
```

Comment:
I believe `SwerveSubsystem::updateOdometry` relies on timestamps and sample data being the same size. I think you may run into array out of bounds errors if that's not the case.

Reply:
i don't believe it does? `updateInputs()` for both the modules and gyro just stream `odometrySamples` straight to an array and the latest timestamp is set to whatever the last value of `odometrySamples` is (so, not reliant on what the size is between loops). outside of those two i don't think it's used

Reply:
also it seemed fine when i ran it on the log

Reply:
The problem is that the `sampleCount` is based on [the minimum number of odometryTimestamps in any module](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java#L381-L382) but `odometryTimestamp` is set [here](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/ModuleIOReal.java#L205-L208) and will be equal to the total number of samples because no filtering goes on. Also, note that `odometryTurnPostions` *does* filter out nulls (probably the right choice) and `odometryDrivePositions` doesn't (which means if you have a sample with no drive positions you're going to have arrays with nulls that you probably don't expect). Because you're doing some filtering, the number of turn positions *can* be smaller than the number of samples. [Right here](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/Module.java#L69) we index turn positions based on the number of samples and may see an out of bounds access. Looks like `updateOdometry` might actually get away with it when calling `getOdometryPositions` because the underlying array is based on sample counts, but it makes the [same mistake](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java#L415) with yaw which is filtered [here](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/GyroIOPigeon2.java#L53-L58).

Just want to be clear that the filtering isn't the underlying problem - you really *have* to do that if you don't know the sample is going to have all signals at all times. The problem is the assumption that the number of timestamps and the number of sample data points will be the same.

Reply:
i'm sorry if i'm still not understanding but i believe `odometryTimestamps` gets it from the `asyncOdometrySamples ` passed into `updateInputs`, [which should be `samplesSince()` in `PhoenixOdometryThread`](https://github.com/HighlanderRobotics/Crescendo/blob/gyro-odo-fix/src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java#L334-L342), which should be the journal which has already been filtered. i think the same goes for turn positions, drive positions, and yaw positions.

do you think we should filter out nulls for `odometryDrivePositions` as well?